source/slang/slang-ir-specialize-function-call.cpp

```diff
@@ -0,0 +1,830 @@
+        // than a declaration. Otherwise there is no
+        // way to generate a specialized callee function.
+        //
+        auto func = as<IRFunc>(call->getCallee());
+        if(!func)
+            return false;
+        if(!func->isDefinition())
+            return false;
+
+        // With the basic checks out of the way, there are
+        // two conditions we care about:
+        //
+        // 1. Should we specialize? This amounts to whether
+        // `func` has any parameters that need specialization.
+        // We will call those "specializable" parameters for
+        // lack of a better name.
+        //
+        // 2. Can we specialize? This amounts to whether the
+        // arguments in `call` that correspond to those
+        // specializable parameters are "suitable" for use
+        // in specialization.
+        //
+        // We are going to answer both of these queries in
+        // a single loop that walks over the parameters of
+        // `func` as well as the arguments to `call`.
+        //
+        // The loop may seem a bit awkward because we are
+        // doing a parallel iteration over a linked list
+        // (the parameters of `func`) and an array (the
+        // arguments of `call`).
+        //
+        bool anySpecializableParam = false;
+        UInt argCounter = 0;
+        for( auto param : func->getParams() )
+        {
+            UInt argIndex = argCounter++;
+            SLANG_ASSERT(argIndex < call->getArgCount());
+            auto arg = call->getArg(argIndex);
+
+            // If the given parameter doesn't need specialization,
+            // then we need to keep looking.
+            //
+            if(!doesParamNeedSpecialization(param))
+                continue;
+
+            // If we have run into a `param` that needs specialization,
+            // then our first condition is met.
+            //
+            anySpecializableParam = true;
+
+            // Now we need to check whether `arg` is actually suitable
+            // for specialization (our second condition). If not, we
+            // can bail out immediately because our second condition
+            // cannot be met.
+            //
+            if(!isArgSuitableForSpecialization(arg))
+                return false;
+        }
+
+        // If we exit the loop, then the second condition must have
+        // been met (all the arguments for specializable parameters
+        // were suitable for specialization), and the result of the
+        // query comes down to the first condition.
+        //
+        return anySpecializableParam;
+    }
+
+    // Of course, now we need to back-fill the predicates that
+    // the above function used to evaluate prameters and arguments.
+
+    bool doesParamNeedSpecialization(IRParam* param)
+    {
+        return condition->doesParamNeedSpecialization(param);
+    }
+
+    bool isArgSuitableForSpecialization(IRInst* inArg)
+    {
```

Comment:
I suppose it might eventually make sense to make the conditions under which an argument is "suitable for specialization" be something a derived pass can specialize.